### PR TITLE
On demand heartbeats via `--heartbeat_on_demand_duration`, used by the tablet throttler

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Fixes https://github.com/vitessio/vitess/issues/10196, and see the general description of the problem and suggested fix in https://github.com/vitessio/vitess/issues/10196.

This PR introduces a new command line flag, `--heartbeat_on_demand_duration`. By default it's zero, which means it's disabled and has no effect (no change in behavior).

When positive, e.g. `--heartbeat_on_demand_duration=5s`, and assuming `--heartbeat_enable` is set, then the tablet does not in fact start generating heartbeats on startup. Instead, it only generates heartbeats on a lease per demand.

In this PR the tablet throttler is currently the only entity which asks the heartbeats writer to generate on demand heartbeats. Right now, the tablet throttler makes such requests _any time_ anyone runs a throttler `check`. Normally, no one does, and so the throttler does not make any request for heartbeats, and so no heartbeats are generated.

However, whenever a `vreplication` workflow begins, or whenever an Online DDL operation begins, the throttler is called (assuming `--enable_lag_throttler` is enabled, of course), and in turn the throttler asks the heartbeat writer to generate heartbeats.

Once the workflow or migration complete, and assuming no other workflows/migrations running, the throttler stops asking for heartbeats, the leas expires, the heartbeat writer ceases to write heartbeats.

The effect is that we keep the binary logs small, as we don't write heartbeats when those are not strictly needed.

As tests go, I've added `--heartbeat_on_demand_duration=5s` to all `endtoend/onlineddl` tests where `--enable_lag_throttler` is set; a couple `endtoend` where `--heartbeat_on_demand_duration=5s` is set are left without  `--heartbeat_on_demand_duration`.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/10196

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/repltracker/writer.go [hunk 1]
```diff
@@ -205,12 +214,42 @@ func (w *heartbeatWriter) recordError(err error) {
...
 }
+
+func (w *heartbeatWriter) RequestHeartbeats() {
+	if w.onDemandDuration == 0 {
+		// heartbeats are not by demand. Therefore they are just coming in on their own (if enabled)
+		return
+	}
+	// In this function we're going to create a timer to activate heartbeats by-demand. Creating a timer has a cost.
+	// Now, this function can be spammed by clients (the lag throttler). We therefore only allow this function to
+	// actually once per X seconds (1/4 of onDemandDuration as a reasonable oversampling value):
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletserver/repltracker/writer.go:231]: ❤️

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

File: go/vt/vttablet/tabletserver/throttle/throttler.go [hunk 2]
```diff
@@ -28,6 +28,7 @@ import (
 	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
 	"vitess.io/vitess/go/vt/topo"
 	"vitess.io/vitess/go/vt/vttablet/tabletserver/connpool"
+	"vitess.io/vitess/go/vt/vttablet/tabletserver/heartbeat"
```
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletserver/throttle/throttler.go:31]: Need to add this to git?

Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/throttle/throttler.go:31]: Whoops! Added.

File: doc/releasenotes/14_0_0_summary.md [hunk 3]
```diff
@@ -88,6 +104,14 @@ API endpoint `/throttler/throttle-app` now accepts a `ratio` query argument, a f
 - `1` means "always throttle"
 - any numbr in between is allowd. For example, `0.3` means "throttle in 0.3 probability", ie on a per request and based on a dice roll, there's a `30%` change a request is denied. Overall we can expect about `30%` of requests to be denied. Example: `/throttler/throttle-app?app=vreplication&ratio=0.25`
 
+#### Heartbeat
+
+The throttler now checks in with the heartbeat writer to request heartbeats, any time it (the throttler) is asked for a check.
+
+When `--heartbeat_on_demand_duration` is not set, there is now change in behavior.
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply doc/releasenotes/14_0_0_summary.md:111]: fixed

File: doc/releasenotes/14_0_0_summary.md [hunk 4]
```diff
@@ -44,14 +44,28 @@ In order to transition, a standalone double-dash (literally, `--`) will cause th
...
+
+`--heartbeat_on_demand_duration` joins the already existing heartbeat flags `--heartbeat_enable` and `--heartbeat_interval` and adds new behavior to heartbeat writes.
+
+`--heartbeat_on_demand_duration` takes a duration value, such as `5s`.
+
+The default value for `--heartbeat_on_demand_duration` is zero, which means the flag is not set and there is no change in behavior.
+
+When `--heartbeat_on_demand_duration` has a positive value, then heartbeats are only injected on demand, per internal requests. For example, when `--heartbeat_on_demand_duration=5s`, the tablet starts without injecting heartbeats. An internal module, like the lag throttle, may request the heartbeat writer for heartbeats. Starting at that point in time, and for the duration of `5s` in our example, the tablet will write heartbeats. If no other requests come in during that duraiton, then the tables then ceases to write heartbeats. If more requests for heartbeats come while heartbeats are being written, then the tablet extends the heartbeat duration for the next `5s` following up each request. Thus, it stops writing heartbeats `5s` after the last request is received.
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply doc/releasenotes/14_0_0_summary.md]: fixed

**Comment by shlomi-noach** [comment]
I think there's a legitimate CI failure in `tabletmanager_throttler` after switching it to `--heartbeat_on_demand_duration`. Will probably not be able to fix this today, and so it's next week for me.

**Comment by shlomi-noach** [comment]
OK, solved the failing test - which actually validated the changes here. Added more testing to validate the changes.

**Comment by shlomi-noach** [comment]
some more legit failures. Will only look at these next week.

**Comment by shlomi-noach** [comment]
Good for a renewed review. Fixed tests. I believe I reduced some flakiness, or at least did not introduce new ones!

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

**Pull request merged**